ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -806,32 +854,98 @@ toCBORAHeaderOrBoundary pm epochSlots abob =
   Mempool integration
 -------------------------------------------------------------------------------}
 
+-- | The hash of a delegation certificate.
+type DlgCertId = Crypto.AbstractHash Blake2b_256 CC.Delegation.Certificate
+
+-- | The hash of an update vote.
+type UpdateVoteId = Crypto.AbstractHash Blake2b_256 CC.Update.Vote
+
+-- | An error type which represents either a UTxO, delegation, update proposal
+-- registration, or update vote error in the Byron era.
+data ByronApplyTxError
+  = ByronApplyTxError !CC.UTxO.UTxOValidationError
+  | ByronApplyDlgError !V.Scheduling.Error
+  | ByronApplyUpdateProposalError !CC.UPI.Error
+  | ByronApplyUpdateVoteError !CC.UPI.Error
```

Comment:
A `CC.UPI.Error` can refer to a few different kinds of update validation errors such as proposal registration and voting errors. Should we expose the more specific errors from ledger?

If not, this could probably just be combined into a single constructor, `ByronApplyUpdateError`.

Reply:
Hmm, ideally, the wrapped error type (`CC.UPI.Error`) only has constructors that can be thrown when applying the respective proposal/error. Given that `CC.UPI.Error` has 13 constructors, that's probably not feasible. I'm okay with the current approach.